iroh/src/baomap/flat.rs

```diff
@@ -0,0 +1,1210 @@
+//! A flat file database implementation.
+//!
+//! This is a simple database implementation that stores all data in the file system.
+//! It is used by the iroh binary.
+use std::collections::{BTreeMap, BTreeSet};
+use std::fmt;
+use std::io::{self, BufReader};
+use std::path::{Path, PathBuf};
+use std::str::FromStr;
+use std::sync::{Arc, RwLock};
+
+use bao_tree::io::outboard::{PostOrderMemOutboard, PreOrderOutboard};
+use bao_tree::io::sync::ReadAt;
+use bao_tree::{blake3, ChunkNum};
+use bao_tree::{BaoTree, ByteNum};
+use bytes::Bytes;
+use futures::future::BoxFuture;
+use futures::future::Either;
+use futures::{Future, FutureExt};
+use iroh_bytes::baomap::{
+    self, ExportMode, ImportMode, ImportProgress, Map, MapEntry, PartialMap, PartialMapEntry,
+    ReadonlyStore, ValidateProgress,
+};
+use iroh_bytes::util::progress::{IdGenerator, ProgressSender};
+use iroh_bytes::{Hash, IROH_BLOCK_SIZE};
+use iroh_io::{AsyncSliceReader, AsyncSliceWriter, File};
+use rand::Rng;
+use range_collections::RangeSet2;
+use tokio::sync::mpsc;
+use tracing::trace_span;
+
+use super::flatten_to_io;
+
```

Comment:
It would be really helpful to have some documentation describing the on disk data structure, and what it relies on